src/common/opt_partition_builder.h

```diff
@@ -204,6 +204,13 @@ class OptPartitionBuilder {
     if (!all_dense) {
       std::vector<size_t>& local_states = states[tid];
       std::vector<bool>& local_default_flags = default_flags[tid];
+      if (row_indices_begin >= row_indices_end) {
```

Comment:
Hey, I saw that you have pushed a fix, presumably found during some local testing.  But really, there's no way I can guess what you are trying to fix.  These few lines look like a hacky workaround to some edge cases.  It's fine, I do that all the time, but can you think of a more general design that doesn't have these conditions?

Reply:
I changed this fix. Thanks for comment!
Seems, currently it doesn't look like some hacky workaround ;)
And I agree that clear and more general design should be proposed for `OptPartitionBuilder`, possibly with strategy patter applying.